### ISSUE TITLE
Ensure balanced mini batches while training

### ISSUE THREAD

**Issue by leonjessen** [rstudio/keras#366]
I am working with an imbalanced binary class data set for classification, with ~90% negative examples and ~10% positive examples and a batch size of 20 when training.

I am interested in ensuring, that each batch used for back-propagating is balanced, such that ~10 data points are sampled from the positive training data and ~10 from the negative. Thereby avoiding that the model is biased towards negative data.

I have been unable to find RStudio/Keras documentation on how to do this? (There is `class_weight` for `fit()`, but I am uncertain if this would achieve my objective)

(Should this not be the proper place to ask this question, please feel free to redirect me)

**Comment by mtoto** [comment]
StackOverflow is probably more appropriate.

**Comment by dfalbel** [comment]
I think class weights will only give different weights to observations in the loss function calculation.
To ensure balanced mini-batches, I think you will need to use a generator that will do this job.

**Comment by leonjessen** [comment]
Thanks @dfalbel - Where might I read more about how to do that?

**Comment by dfalbel** [comment]
[Here](https://keras.io/models/model/#methods) at Keras docs there's some explanation on `class_weights` parameter. Quoting the relevant part below:

>  * class_weight: Optional dictionary mapping class indices (integers) to a weight (float) value, used for weighting the loss function (during training only). This can be useful to tell the model to "pay more attention" to samples from an under-represented class.
> * sample_weight: Optional Numpy array of weights for the training samples, used for weighting the loss function (during training only). You can either pass a flat (1D) Numpy array with the same length as the input samples (1:1 mapping between weights and samples), or in the case of temporal data, you can pass a 2D array with shape (samples, sequence_length), to apply a different weight to every timestep of every sample. In this case you should make sure to specify sample_weight_mode="temporal" in compile().

For the generator, I would write something like (assuming `pos` and `neg` are lists with 2 elements `x`  - a matrix of features - and `y` - a vector of responses (0, 1).

```
balanced_generator <- function(pos, neg, batch_size) {
  
  id_pos <- NULL
  id_neg <- NULL
  
  function() {
    
    if(length(id_pos) < batch_size/2) id_pos <- nrow(pos$x)
    if(length(id_neg) < batch_size/2) id_neg <- nrow(neg$x)
    
    s_id_pos <- sample(id_pos, size = batch_size/2)
    s_id_neg <- sample(id_neg, size = batch_size/2)
    
    id_pos <<- id_pos[!id_pos %in% s_id_pos]
    id_neg <<- id_neg[!id_neg %in% s_id_neg]
    
    
    list(
     rbind(pos$x[s_id_pos,], neg$x[s_id_neg,]),
     c(pos$y[s_id_pos], pos$y[s_id_neg])
    )
  }
  
}
```

This way we are generating batch that have batch_size/2 observations for each class.
More info on generators [here](https://keras.rstudio.com/articles/faq.html#how-can-i-use-keras-with-datasets-that-dont-fit-in-memory)

**Comment by leonjessen** [comment]
Awesome @dfalbel - This looks like just the thing I was looking for! Thanks!

**Comment by leonjessen** [comment]
So @dfalbel , I slightly modfied your suggestion and my solution ended up being like so:
``` r
balanced_generator = function(X_data, Y_data, batch_size){
  function(){
    i_0 = sample(x = which(Y_data == 0), size = batch_size / 2, replace = TRUE)
    i_1 = sample(x = which(Y_data == 1), size = batch_size / 2, replace = TRUE)
    i   = c(rbind(i_0, i_1))
    list(X_data[i,], Y_data[i])
  }
}
```
and use the `fit_generator()` function to train the network

**Comment by dfalbel** [comment]
Great! Note that you will have the overhead of filtering X and Y for each batch, so it could slow down your fitting, if the dataset is too large. 

The ideal would be to implement this inside the graph, by using tfdatasets and tf$ manip functions, so it could run multithreaded, but I don't see a simple way to do this using tf primitives.
